<a href="https://colab.research.google.com/github/GonorAndres/Analisis_Numerico_2025-2/blob/main/Practica3/Ejercicio23.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
from numpy.linalg import cond, norm
from math import acos, degrees

import pandas as pd

def resolver_sistema_por_gauss(A, b):
    """Resuelve Ax = b usando eliminación de Gauss sin pivoteo."""
    A = A.astype(float)
    b = b.astype(float)
    n = len(b)

    for i in range(n):
        for j in range(i + 1, n):
            factor = A[j, i] / A[i, i]
            A[j, i:] -= factor * A[i, i:]
            b[j] -= factor * b[i]

    x = np.zeros(n)
    for i in reversed(range(n)):
        x[i] = (b[i] - np.dot(A[i, i+1:], x[i+1:])) / A[i, i]

    return x

def minimos_cuadrados_manual(A, b):
    AtA = A.T @ A
    Atb = A.T @ b
    return resolver_sistema_por_gauss(AtA.copy(), Atb.copy())

def condicion_matriz_linalg(A):
    return cond(A, 2)

In [4]:
# Para el ejercicio 23a: Resolver sistema de mínimos cuadrados con cualquier método.
# Usamos el mismo método manual por ecuaciones normales ya implementado en el ejercicio anterior

# Datos del inciso 23a
A_23a = np.array([
    [0.16, 0.10],
    [0.17, 0.11],
    [2.02, 1.29]
])

b_23a = np.array([0.26, 0.28, 3.31])

# Solución usando ecuaciones normales (manualmente)
x_23a = minimos_cuadrados_manual(A_23a, b_23a)
b_tilde_23a = A_23a @ x_23a
residuo_23a = norm(b_23a - b_tilde_23a)

# Datos del inciso 23b (con perturbación)
b_23b = np.array([0.27, 0.25, 3.33])
x_23b = minimos_cuadrados_manual(A_23a, b_23b)
b_tilde_23b = A_23a @ x_23b
residuo_23b = norm(b_23b - b_tilde_23b)

# Comparación
df_23 = pd.DataFrame({
    "Inciso": ["23a (original)", "23b (perturbado)"],
    "x̃ obtenida (redondeada)": [np.round(x_23a, 6), np.round(x_23b, 6)],
    "Norma del residuo": [residuo_23a, residuo_23b]
})

df_23

Inciso x̃ obtenida (redondeada)  Norma del residuo
0    23a (original)               [1.0, 1.0]       3.405189e-13
1  23b (perturbado)    [7.008887, -8.395663]       2.168268e-02

### Análisis del ejercicio 23

- En el caso **original**, la solución obtenida fue:

  \[
  $\tilde{x}$ =
  \begin{bmatrix}
  1 \\
  1
  \end{bmatrix},
  \]

  con un residuo prácticamente nulo, lo que indica un ajuste perfecto del modelo a los datos.

- En el caso **perturbado**, una **pequeña variación** en el vector $ \bar{b}$ condujo a una solución:

  \[
  $\tilde{x}$ $\approx$
  \begin{bmatrix}
  7.008887 \\
  -8.395663
  \end{bmatrix},
  \]

  junto con un residuo considerablemente mayor.

---

Esto muestra que el sistema es **sensible a perturbaciones** en los datos, lo cual sugiere que está **mal condicionado**. Incluso cambios pequeños en el vector $ \bar{b} $ pueden causar variaciones drásticas en la solución $ \tilde{x} $.

